### ISSUE TITLE
`<expected>`: Investigate reusing tail padding of the union

### ISSUE THREAD

**Issue by frederick-vs-ja** [microsoft/STL#4131]
Given `[[msvc::no_unique_address]]` landed in Clang 18, we can applied it to union members (possibly with the whole union) in `expected`.

E.g. in this example
```C++
struct IC {
    int i;
    char c;
};
std::expected<IC, int> e;
```
its possible to make `sizeof(e) == 8` because the size of the union is 8, while the first 5 bytes are value bytes, and `bool` flag can live in the tail padding of the union.

We need to ensure that the tail padding won't be furtherly reused - otherwise, when the `expected` object is itself potentially-overlapping, reconstruction of a union member may overwrite the data not controlled by the implementation.

See also LLVM-68552, LLVM-68733, and LLVM-69673.

**Comment by StephanTLavavej** [comment]
Related to #1364. That issue is vNext because C++20 is locked down, but we can do whatever we want for C++23 until `/std:c++23` is finalized. Avoiding the tail-padding-reuse problem is definitely very important, thanks for mentioning it and linking libc++'s bug!

**Comment by frederick-vs-ja** [comment]
Something isn't working well with Windows/MSVC ABI. [Godbolt link](https://godbolt.org/z/PaWz83Pxh)

Is it possible to work around the situation?

```C++
#if __has_cpp_attribute(no_unique_address)
#define NO_UNIQUE_ADDRESS [[no_unique_address]]
#elif __has_cpp_attribute(msvc::no_unique_address)
#define NO_UNIQUE_ADDRESS [[msvc::no_unique_address]]
#else
#error [[no_unique_address]] and [[msvc::no_unique_address]] are ABI-critical!
#endif

int main() {
    struct X {
        // workaround for missing of CWG-43 resolution
        X() = default;
        constexpr explicit X(int x) noexcept : n{x}, c{} {}
        constexpr X(int x, char y) noexcept : n{x}, c{y} {}

        int n;
        char c;
    };
    struct Y {
        NO_UNIQUE_ADDRESS X x;
        char d;
    };
    static_assert(sizeof(X) == 2 * sizeof(int));
    static_assert(sizeof(Y) == sizeof(X));
}
```

**Comment by CaseyCarter** [comment]
If you mean "clang and MSVC don't agree", [they do when both using MSABI](https://godbolt.org/z/cKh9KeK9v). If you mean "I wish MSVCABI would store `d` in the tail padding of `X`" [that seems intended](https://godbolt.org/z/nEna9bK3z). I'm not sure _why_ it's intended, but can poke the layout guru on the compiler team if it's important to you.

**Comment by frederick-vs-ja** [comment]
> If you mean "clang and MSVC don't agree", [they do when both using MSABI](https://godbolt.org/z/cKh9KeK9v).

No, I've verified that they agreed. But MSABI doesn't seem to support my original intent - mirroring the implementation strategy of libc++'s [`expected`](https://github.com/llvm/llvm-project/blob/43f044ba7ea2cd668805c01055eb2dff70532ba3/libcxx/include/__expected/expected.h) with tail padding bytes reused.

> If you mean "I wish MSVCABI would store `d` in the tail padding of `X`" [that seems intended](https://godbolt.org/z/nEna9bK3z). I'm not sure _why_ it's intended, but I can poke the layout guru on the compiler team if it's important to you.

Consider `std::expected<T, E>` where both `T` and `E` have at least 1 tail padding byte. What I desired is storing the `bool` `_Has_value` flag into the tail padding, which already is achieved by libc++'s `expected` on Itanium ABI.

It requires some additional works to ensure that reconstruction of `T` or `E` doesn't ruin the tail padding bytes of the whole `expected`. The works were done by the posted LLVM-68733 and LLVM-69673 in libc++.
